core/shared/src/main/scala/it/unibo/scafi/lib/Gradients.scala

```diff
@@ -5,26 +5,269 @@
+                     commRadius: Double = 0.2,
+                     lagMetric: => Double = nbrLag().toMillis): Double = {
+      val avgFireInterval = meanCounter(deltaTime().toMillis, 1000000)
+      val speed = 1.0 / avgFireInterval
+
+      rep((Double.PositiveInfinity, Double.PositiveInfinity)) { case (spatialDist: Double, tempDist: Double) =>
+        mux(source) {
+          (0.0, 0.0)
+        } {
+          minHoodPlus {
+            val newEstimate = Math.max(nbr {
+              spatialDist
+            } + metric(), speed * nbr {
+              tempDist
+            } - commRadius)
+            (newEstimate, nbr {
+              tempDist
+            } + lagMetric / 1000.0)
+          }
+        }
+      }._1
+    }
+
+    def crfGradient(
+                     source: Boolean,
+                     metric: Metric = nbrRange,
+                     raisingSpeed: Double = 5,
+                     lagMetric: => Double = nbrLag().toMillis): Double =
+      rep((Double.PositiveInfinity, 0.0)) {
+        case (g, speed) =>
+          mux(source){ (0.0, 0.0) }{
+            implicit def durationToDouble(fd: FiniteDuration): Double = fd.toMillis.toDouble / 1000.0
+            case class Constraint(nbr: ID, gradient: Double, nbrDistance: Double)
+
+            val constraints = foldhoodPlus[List[Constraint]](List.empty)(_ ++ _){
+              val (nbrg, d) = (nbr{g}, metric())
+              mux(nbrg + d + speed * lagMetric <= g){ List(Constraint(nbr{mid()}, nbrg, d)) }{ List() }
+            }
+
+            if(constraints.isEmpty){
+              (g + raisingSpeed * deltaTime(), raisingSpeed)
+            } else {
+              (constraints.map(c => c.gradient + c.nbrDistance).min, 0.0)
+            }
+          }
+      }._1
+
+    /**
+      * Idea: a device should change its estimate only for significant errors.
+      * Useful when devices far from the source need only coarse estimates.
+      * Flex gradient provides tunable trade-off between precision and communication cost.
+      *
+      * @param source Source fields of devices from which the gradient is calculated
+      * @param epsilon Parameter expressing tolerance wrt changes
+      * @param delta Distortion into the distance measure, such that neighbor distance is
+      *              never considered to be less than delta * communicationRadius.
+      * @param communicationRadius
+      * @return
+    */
+    def flexGradient(source: Boolean,
+                     metric: Metric = nbrRange,
+                     epsilon: Double = 0.5,
+                     delta: Double = 1.0,
+                     communicationRadius: Double = 1.0
+            ): Double =
+      rep(Double.PositiveInfinity){ g =>
+        def distance = Math.max(nbrRange(), delta * communicationRadius)
+
+        import Builtins.Bounded._ // for min/maximizing over tuples
+        val maxLocalSlope: (Double,ID,Double,Double) = //??? // TODO: typeclass resolution for tuple (Double,ID,Double,Double) broke
```

Comment:
Consider adding an implicit parameter to the method for dealing with this (and remove the comment).
That is, we could move this requirement (of providing the typeclass instance of `(Double,ID,Double,Double)`) from the library to the user of the library..